pkg/clients/bigquery.go

```diff
@@ -0,0 +1,239 @@
+}
+
+// Write a failure event entry if there is a failure in processing the event.
+// This is used by the webhook service.
+func (bq *BigQuery) WriteFailureEvent(ctx context.Context, failureEventTableID, deliveryID, createdAt string) error {
+	inserter := bq.client.Dataset(bq.datasetID).Table(failureEventTableID).Inserter()
+	items := []*FailureEventEntry{
+		// FailureEventEntry implements the ValueSaver interface.
+		{deliveryID: deliveryID, createdAt: createdAt},
+	}
+	if err := inserter.Put(ctx, items); err != nil {
+		return fmt.Errorf("failed to execute WriteFailureEvent for deliveryID %s: %w", deliveryID, err)
+	}
+
+	return nil
+}
+
+// Retrieve the latest checkpoint cursor value (deliveryID) in the checkpoint
+// table. This is used by the retry service.
+func (bq *BigQuery) RetrieveCheckpointID(ctx context.Context, checkpointTableID string) (string, error) {
+	// Construct a query.
+	q := bq.client.Query("SELECT delivery_id FROM `@projectID.@datasetID.@checkpointTableID` ORDER BY created DESC LIMIT 1")
+
+	// Set the params for the query
+	q.Parameters = []bigquery.QueryParameter{
+		{
+			Name:  "projectID",
+			Value: bq.projectID,
+		},
+		{
+			Name:  "datasetID",
+			Value: bq.datasetID,
+		},
+		{
+			Name:  "checkpointTableID",
+			Value: checkpointTableID,
+		},
+	}
+
+	// Execute the query.
+	res, err := q.Read(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to make read request to BigQuery: %w", err)
+	}
+
+	var rows []bigquery.Value
+	nextErr := res.Next(&rows)
+
+	if nextErr != nil {
+		// if no checkpoint ID exists, return the empty string with no error
+		if errors.Is(nextErr, iterator.Done) {
+			return "", nil
+		}
+		return "", fmt.Errorf("failed to iterate over query response: %w", nextErr)
+	}
+
+	if len(rows) == 0 {
+		return "", fmt.Errorf("unexpected response from RetrieveCheckpointID : %s", rows)
+	}
+
+	checkpoint, ok := rows[0].(string)
+	if !ok {
+		return "", fmt.Errorf("failed to convert row value %v to string: (got %T)", rows[0], rows[0])
+	}
+
+	return checkpoint, nil
+}
+
+// Write the latest checkpoint that was successfully processed.
+// This is used by the retry service.
+func (bq *BigQuery) WriteCheckpointID(ctx context.Context, checkpointTableID, deliveryID, createdAt string) error {
+	inserter := bq.client.Dataset(bq.datasetID).Table(checkpointTableID).Inserter()
+	items := []*CheckpointEntry{
+		// CheckpointEntry implements the ValueSaver interface.
+		{deliveryID: deliveryID, createdAt: createdAt},
+	}
+	if err := inserter.Put(ctx, items); err != nil {
+		return fmt.Errorf("failed to execute WriteCheckpointID for deliveryID %s: %w", deliveryID, err)
+	}
+
+	return nil
+}
+
+// Helper method to execute a count query for a given table by deliveryID and
+// return the count.
+func (bq *BigQuery) makeCountQuery(ctx context.Context, tableID, deliveryID string) (int, error) {
+	q := bq.client.Query("SELECT COUNT(1) FROM `@projectID.@datasetID.@tableID` WHERE delivery_id = @deliveryID")
```

Comment:
I moved this into this helper method because with the exception of the table name, the two calling methods are nearly identical.